Add schema files for request validation.

jbasden opened the pull request:
This commit adds the JSON schema for log_events and metric_datum.
It also adds the core logic for schema.py.

Commit:
Add schema files for request validation.

This commit adds the JSON schema for log_events and metric_datum.
It also adds the core logic for schema.py.

Commit:
Fix json formatting.

Took out backslashes left over from python implementation.
Also removed two files that were solely for experimentation purposes.

Commit:
Fix list formatting.

Indent everything with tabs instead of spaces.

Commit:
Fix json formatting using jsonlint.

Ran two files, log_event.json and metric_datum.json through the linter.

ylynn commented on metricpublisher/schemas/log_event.json:
```diff
@@ -0,0 +1,14 @@
+    "type": "object",
+    "properties": {
+        "request_id": {
+            "type": "number"
```
request_id should be a string

jbasden replied:
Made the change.

ylynn commented on metricpublisher/schemas/metric_datum.json:
```diff
@@ -0,0 +1,98 @@
+            "type": "number"
+        }
+    },
+    "required": [
```
Did you try these two cases with cli: 1. both Value and StatisticValues exist 2. both Value and StatisticValues don't exist?

jbasden replied:
Yes - both of them give an error. I was unaware that they were supposed to be collectively exhaustive and mutually exclusive. Should this be reflected in the JSON schema?

ylynn replied:
Yes.

Commit:
Change metric_datum.json to reflect mutually exclusive and collectively exhaustive events.

"Value" and "StatisticValues" are mutually exclusive and collectively exhaustive.
This commit adds a "oneOf" statement to the metric_datum.json JSONSchema to reflect this.
It also changes request id from a number to a string.

ylynn reviewed (changes requested):
Include unit tests for validate function, otherwise you cannot know if they will work.

Commit:
Add unit tests for schema validation.

This commit adds ten tests for validating whether or not
a user's given input is valid against log_event.json. It
creates the main logic to allow these tests to run automatically.

ylynn commented:
Describe what test you ran?

Commit:
Add docstrings to test handlers.

This commit adds a quick explanation (in the docstring)
of each of the tests ran. The tests in the test_handler
file can be run by executing the makefile in the root directory. (make)

ylynn commented:
I was asking you to describe what test you ran in your pull request description. Like I said before, you can just say "run make to test" or "make passed".

ylynn reviewed (changes requested):
Add tests for invalid "Dimension" value.

ylynn commented on metricpublisher/schemas/log_event.json:
```diff
@@ -0,0 +1,129 @@
+        "request_id": {
+            "type": "string"
+        },
+        "metric_data": {
```
For the attributes that have properties under it, have definition files for them separately and then ref them. See example https://github.com/awslabs/aws-serverless-dynamodb-event-sourcing/tree/master/app/schemas

ylynn replied:
For the attributes that have properties under it, have definition files for them separately and then ref them. See example https://github.com/awslabs/aws-serverless-dynamodb-event-sourcing/tree/master/app/schemas

Commit:
Add test_dimesion_type_invalid.

This commit adds a test to make sure that the input type of 'Dimension'
is valid. It asserts that an error is thrown when another data type besides
array (python list) is input (set, in this example). All tests passed by
executing the makefile (make).

ylynn commented on metricpublisher/schemas/log_event.json:
```diff
@@ -0,0 +1,129 @@
+                    },
+                    "Dimensions": {
+                        "type": "array",
+                        "properties": {
```
This is wrong. For type array, you should have something like this: 
```
"type": "array",
  "items": {
    "type": "object",
    "properties": {
```
So the other problem is that your tests did not help catch this.

Commit:
Refactor log_event.json to improve reusibility of JSON schema.

This commit defines a "definitions" section with all of the
JSON schemas used in request validation. make passed.

Commit:
Remove single line from schema.py.

This commit removes a line that was inserted solely for
experimentation purposes.

ylynn commented on test/unit/events.py:
```diff
@@ -0,0 +1,313 @@
+    }
+
+@pytest.fixture()
+def Dimension_type_invalid():
```
Dimensions invalid cases include: 
1. not an array
2. wrong type of item in the array
3. wrong property of item

jbasden replied:
Dimension_type_invalid deals with 1. Will add separate tests for 2 and 3.

Commit:
Add 2 tests for schema validation.

This commit adds 2 new schema validation unit tests.
The first ensures the type of each item in the dimensions
array is valid. The second ensures that the required properties
are present in each array item.

ylynn commented on metricpublisher/schemas/log_event.json:
```diff
@@ -0,0 +1,162 @@
+    "title": "Schema for the log_event lambda request object",
+    "type": "object",
+    "definitions": {
+        "request_id": {
```
Only include the attributes have "items" or "properties" under them in "definitions". Simple ones like "request_id" does not need to have separate definitions.

jbasden replied:
Oh got ya. I thought you meant to have separate definitions for the properties themselves, not the ones that have properties underneath them...

Commit:
Remove JSON references to simple schema.

This commit removes all references that do not have
"items" or "properties" under them. make passed.

ylynn commented on metricpublisher/schemas/log_event.json:
```diff
@@ -0,0 +1,149 @@
+                "Value"
+            ]
+        },
+        "Dimensions": {
```
You have some attributes capitalized ("Dimensions" "StatisticValues" "SampleCount" ...) and some not ("metric_datum" "metric_data"). Make them all lower case to be consistent.

Commit:
Convert camelcase to lowercase with underscores.

This commit fixes the names in schema validation so
that everything is consistent (lower case with underscores).
make passed.

honglu commented on metricpublisher/lambda_handler.py:
```diff
@@ -1 +1,35 @@
+    """Log event.
+
+    Parameters:
+        event (dict): the user input
```
"the user input" is not very helpful in the doc. Better to more descriptive by mentioning it is the metric information user would like to emit.
